relay-cardinality/src/redis.rs

```diff
@@ -341,6 +418,250 @@ impl SlidingWindow {
+enum CacheUpdate<'a> {
+    Empty,
+    Update {
+        inner: RwLockWriteGuard<'a, CacheInner>,
+        key: CacheKeyRef<'a>,
+    },
+}
+
+impl<'a> CacheUpdate<'a> {
+    /// Marks a hash as accepted in the cache, future checks of the item will immediately mark the
+    /// item as accepted.
+    pub fn accept(&mut self, hash: u32) {
+        if let Self::Update { inner, key } = self {
+            inner.cache.entry_ref(key).or_default().0.insert(hash);
+        }
+    }
+}
+
+/// Critical section of the [`Cache`].
+#[derive(Debug, Default)]
+struct CacheInner {
+    cache: hashbrown::HashMap<CacheKey, CacheValue>,
+    current_window: u64,
+}
+
+/// Scope specific information of the cache.
+#[derive(Debug, Default)]
+struct CacheValue(hashbrown::HashSet<u32>);
+
+impl CacheValue {
+    fn check(&self, hash: u32, limit: usize) -> CacheOutcome {
+        if self.0.contains(&hash) {
+            // Local cache copy contains the hash -> accept it straight away
+            CacheOutcome::Accepted
+        } else if self.0.len() >= limit {
+            // We have more or the same amount of items in the local cache as the cardinality
+            // limit -> this new item/hash is rejected.
+            CacheOutcome::Rejected
+        } else {
+            // Check with Redis.
+            CacheOutcome::Unknown
+        }
+    }
+}
+
+/// Key/scoping for the cardinality cache.
+#[derive(Debug, PartialEq, Eq)]
+struct CacheKey {
+    organization_id: OrganizationId,
+    item_scope: String,
+}
+
+impl Hash for CacheKey {
+    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
+        CacheKeyRef {
+            organization_id: self.organization_id,
+            item_scope: &self.item_scope,
+        }
+        .hash(state)
+    }
+}
+
+impl hashbrown::Equivalent<CacheKeyRef<'_>> for CacheKey {
+    fn equivalent(&self, key: &CacheKeyRef<'_>) -> bool {
+        let CacheKeyRef {
+            organization_id,
+            item_scope,
+        } = key;
+
+        self.organization_id == *organization_id && &self.item_scope == item_scope
+    }
+}
+
+impl From<&CacheKeyRef<'_>> for CacheKey {
+    fn from(value: &CacheKeyRef<'_>) -> Self {
+        Self {
+            organization_id: value.organization_id,
+            item_scope: value.item_scope.to_owned(),
+        }
+    }
+}
+
+/// A borrowed [`CacheKey`].
+#[derive(Debug, PartialEq, Eq, Hash)]
+struct CacheKeyRef<'a> {
+    organization_id: OrganizationId,
+    item_scope: &'a str,
+}
+
+impl hashbrown::Equivalent<CacheKey> for CacheKeyRef<'_> {
+    fn equivalent(&self, key: &CacheKey) -> bool {
+        let CacheKey {
+            organization_id,
+            item_scope,
+        } = key;
+
+        self.organization_id == *organization_id && self.item_scope == item_scope
+    }
+}
+
+/// Helper to aggregate cache hits by scope before reporting them to statsd.
+#[derive(Debug)]
+struct HitCounter(hashbrown::HashMap<String, (i64, i64)>);
```

Comment:
What's the reason for using a non-standard map implementation here?

Reply:
Needed for `entry_ref`, the alternative is pretty awkward (`get_mut` followed by an insert if `None`)